File: 17_Bullfight/python/bullfight.py
import math
import random
from typing import Dict, List, Literal, Tuple, Union


def print_n_newlines(n: int) -> None:
    for _ in range(n):
        print()


def determine_player_kills(
    bull_quality: int,
    player_type: Literal["TOREAD", "PICADO"],
    plural_form: Literal["ORES", "RES"],
    job_qualities: List[str],
) -> float:
    bull_performance = 3 / bull_quality * random.random()
    if bull_performance < 0.37:
        job_quality_factor = 0.5
    elif bull_performance < 0.5:
        job_quality_factor = 0.4
    elif bull_performance < 0.63:
        job_quality_factor = 0.3
    elif bull_performance < 0.87:
        job_quality_factor = 0.2
    else:
        job_quality_factor = 0.1
    job_quality = math.floor(10 * job_quality_factor + 0.2)  # higher is better
    print(f"THE {player_type}{plural_form} DID A {job_qualities[job_quality]} JOB.")
    if job_quality >= 4:
        if job_quality == 5:
            player_was_killed = random.choice([True, False])
            if player_was_killed:
                print(f"ONE OF THE {player_type}{plural_form} WAS KILLED.")
            elif player_was_killed:
                print(f"NO {player_type}{plural_form} WERE KILLED.")
        else:
            if player_type != "TOREAD":
                killed_horses = random.randint(1, 2)
                print(
                    f"{killed_horses} OF THE HORSES OF THE {player_type}{plural_form} KILLED."
                )
            killed_players = random.randint(1, 2)
            print(f"{killed_players} OF THE {player_type}{plural_form} KILLED.")
    print()
    return job_quality_factor


def calculate_final_score(
    move_risk_sum: float, job_quality_by_round: Dict[int, float], bull_quality: int
) -> float:
    quality = (
        4.5
        + move_risk_sum / 6
        - (job_quality_by_round[1] + job_quality_by_round[2]) * 2.5
        + 4 * job_quality_by_round[4]
        + 2 * job_quality_by_round[5]
        - (job_quality_by_round[3] ** 2) / 120
        - bull_quality
    ) * random.random()
    if quality < 2.4:
        return 0
    elif quality < 4.9:
        return 1
    elif quality < 7.4:
        return 2
    else:
        return 3


def print_header() -> None:
    print(" " * 34 + "BULL")
    print(" " * 15 + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY")
    print_n_newlines(2)


def print_instructions() -> None:
    print("HELLO, ALL YOU BLOODLOVERS AND AFICIONADOS.")
    print("HERE IS YOUR BIG CHANCE TO KILL A BULL.")
    print()
    print("ON EACH PASS OF THE BULL, YOU MAY TRY")
    print("0 - VERONICA (DANGEROUS INSIDE MOVE OF THE CAPE)")
    print("1 - LESS DANGEROUS OUTSIDE MOVE OF THE CAPE")
    print("2 - ORDINARY SWIRL OF THE CAPE.")
    print()
    print("INSTEAD OF THE ABOVE, YOU MAY TRY TO KILL THE BULL")
    print("ON ANY TURN: 4 (OVER THE HORNS), 5 (IN THE CHEST).")
    print("BUT IF I WERE YOU,")
    print("I WOULDN'T TRY IT BEFORE THE SEVENTH PASS.")
    print()
    print("THE CROWD WILL DETERMINE WHAT AWARD YOU DESERVE")
    print("(POSTHUMOUSLY IF NECESSARY).")
    print("THE BRAVER YOU ARE, THE BETTER THE AWARD YOU RECEIVE.")
    print()
    print("THE BETTER THE JOB THE PICADORES AND TOREADORES DO,")
    print("THE BETTER YOUR CHANCES ARE.")


def print_intro() -> None:
    print_header()
    want_instructions = input("DO YOU WANT INSTRUCTIONS? ")
    if want_instructions != "NO":
        print_instructions()
    print_n_newlines(2)


def ask_bool(prompt: str) -> bool:
    while True:
        answer = input(prompt).lower()
        if answer == "yes":
            return True
        elif answer == "no":
            return False
        else:
            print("INCORRECT ANSWER - - PLEASE TYPE 'YES' OR 'NO'.")


def ask_int() -> int:
    while True:
        foo = float(input())
        if foo != float(int(abs(foo))):  # we actually want an integer
            print("DON'T PANIC, YOU IDIOT!  PUT DOWN A CORRECT NUMBER")
        elif foo < 3:
            break
    return int(foo)


def did_bull_hit(
    bull_quality: int,
    cape_move: int,
    job_quality_by_round: Dict[int, float],
    move_risk_sum: float,
) -> Tuple[bool, float]:
    # The bull quality is a grade: The lower the grade, the better the bull
    if cape_move == 0:
        move_risk: Union[int, float] = 3
    elif cape_move == 1:
        move_risk = 2
    else:
        move_risk = 0.5
    move_risk_sum += move_risk
    bull_strength = 6 - bull_quality
    bull_hit_factor = (  # the higher the factor, the more "likely" it hits
        (bull_strength + move_risk / 10)
        * random.random()
        / (
            (
                job_quality_by_round[1]
                + job_quality_by_round[2]
                + job_quality_by_round[3] / 10
            )
            * 5
        )
    )
    bull_hit = bull_hit_factor >= 0.51
    return bull_hit, move_risk_sum


def handle_bullkill_attempt(
    kill_method: int,
    job_quality_by_round: Dict[int, float],
    bull_quality: int,
    gore: int,
) -> int:
    if kill_method not in [4, 5]:
        print("YOU PANICKED.  THE BULL GORED YOU.")
        gore = 2
    else:
        bull_strength = 6 - bull_quality
        kill_probability = (
            bull_strength
            * 10
            * random.random()
            / (
                (job_quality_by_round[1] + job_quality_by_round[2])
                * 5
                * job_quality_by_round[3]
            )
        )
        if kill_method == 4:
            if kill_probability > 0.8:
                gore = 1
        else:
            if kill_probability > 0.2:
                gore = 1
        if gore == 0:
            print("YOU KILLED THE BULL!")
            job_quality_by_round[5] = 2
            return gore
    return gore


def final_message(
    job_quality_by_round: Dict[int, float], bull_quality: int, move_risk_sum: float
) -> None:
    print_n_newlines(3)
    if job_quality_by_round[4] == 0:
        print("THE CROWD BOOS FOR TEN MINUTES.  IF YOU EVER DARE TO SHOW")
        print("YOUR FACE IN A RING AGAIN, THEY SWEAR THEY WILL KILL YOU--")
        print("UNLESS THE BULL DOES FIRST.")
    else:
        if job_quality_by_round[4] == 2:
            print("THE CROWD CHEERS WILDLY!")
        elif job_quality_by_round[5] == 2:
            print("THE CROWD CHEERS!")
            print()
        print("THE CROWD AWARDS YOU")
        score = calculate_final_score(move_risk_sum, job_quality_by_round, bull_quality)
        if score == 0:
            print("NOTHING AT ALL.")
        elif score == 1:
            print("ONE EAR OF THE BULL.")
        elif score == 2:
            print("BOTH EARS OF THE BULL!")
            print("OLE!")
        else:
            print("OLE!  YOU ARE 'MUY HOMBRE'!! OLE!  OLE!")
        print()
        print("ADIOS")
        print_n_newlines(3)


def main() -> None:
    print_intro()
    move_risk_sum: float = 1
    job_quality_by_round: Dict[int, float] = {4: 1, 5: 1}
    job_quality = ["", "SUPERB", "GOOD", "FAIR", "POOR", "AWFUL"]
    bull_quality = random.randint(
        1, 5
    )  # the lower the number, the more powerful the bull
    print(f"YOU HAVE DRAWN A {job_quality[bull_quality]} BULL.")
    if bull_quality > 4:
        print("YOU'RE LUCKY.")
    elif bull_quality < 2:
        print("GOOD LUCK.  YOU'LL NEED IT.")
        print()
    print()

    # Round 1: Run Picadores
    player_type: Literal["TOREAD", "PICADO"] = "PICADO"
    plural_form: Literal["ORES", "RES"] = "RES"
    job_quality_factor = determine_player_kills(
        bull_quality, player_type, plural_form, job_quality
    )
    job_quality_by_round[1] = job_quality_factor

    # Round 2: Run Toreadores
    player_type = "TOREAD"
    plural_form = "ORES"
    determine_player_kills(bull_quality, player_type, plural_form, job_quality)
    job_quality_by_round[2] = job_quality_factor
    print_n_newlines(2)

    # Round 3
    job_quality_by_round[3] = 0
    while True:
        job_quality_by_round[3] = job_quality_by_round[3] + 1
        print(f"PASS NUMBER {job_quality_by_round[3]}")
        if job_quality_by_round[3] >= 3:
            run_from_ring = ask_bool("HERE COMES THE BULL.  TRY FOR A KILL? ")
            if not run_from_ring:
                print("CAPE MOVE? ", end="")
        else:
            print("THE BULL IS CHARGING AT YOU!  YOU ARE THE MATADOR--")
            run_from_ring = ask_bool("DO YOU WANT TO KILL THE BULL? ")
            if not run_from_ring:
                print("WHAT MOVE DO YOU MAKE WITH THE CAPE? ", end="")
        gore = 0
        if not run_from_ring:
            cape_move = ask_int()
            bull_hit, move_risk_sum = did_bull_hit(
                bull_quality, cape_move, job_quality_by_round, move_risk_sum
            )
            if bull_hit:
                gore = 1
            else:
                continue
        else:
            print()
            print("IT IS THE MOMENT OF TRUTH.")
            print()
            kill_method = int(input("HOW DO YOU TRY TO KILL THE BULL? "))
            gore = handle_bullkill_attempt(
                kill_method, job_quality_by_round, bull_quality, gore
            )
            if gore == 0:
                break
        if gore > 0:
            if gore == 1:
                print("THE BULL HAS GORED YOU!")
            death = False
            while True:
                if random.randint(1, 2) == 1:
                    print("YOU ARE DEAD.")
                    job_quality_by_round[4] = 1.5
                    death = True
                    break
                else:
                    print("YOU ARE STILL ALIVE.")
                    print()
                    print("DO YOU RUN FROM THE RING? ", end="")
                    run_from_ring = ask_bool("DO YOU RUN FROM THE RING? ")
                    if not run_from_ring:
                        print("YOU ARE BRAVE.  STUPID, BUT BRAVE.")
                        if random.randint(1, 2) == 1:
                            job_quality_by_round[4] = 2
                            death = True
                            break
                        else:
                            print("YOU ARE GORED AGAIN!")
                    else:
                        print("COWARD")
                        job_quality_by_round[4] = 0
                        death = True
                        break

            if death:
                break

    final_message(job_quality_by_round, bull_quality, move_risk_sum)


if __name__ == "__main__":
    main()

File: 17_Bullfight/java/Bullfight.java
import java.util.Random;
import java.util.Scanner;

/**
 * BULLFIGHT
 * <p>
 * Converted from BASIC to Java by Aldrin Misquitta (@aldrinm)
 */
public class Bullfight {

    private static final int MAX_PASSES_BEFORE_CHARGE = 3;

    public static void main(String[] args) {
        printHeader();

        Scanner scanner = new Scanner(System.in);
        System.out.print("\n\n");

        var instructionsChoice = readInstructionsChoice(scanner);
        if (instructionsChoice) {
            printInstructions();
        }

        Random random = new Random();

        //initialize the game with default values
        GameState gameState = new GameState();

        //Randomly select a bull grade
        int randomGrade = (int) (random.nextFloat() * BullGrade.values().length + 1);
        var bullGrade = BullGrade.fromValue(randomGrade);

        printBullGradeInfo(bullGrade);

        System.out.println();

        //D[1] in the original
        gameState.picadoresDamage = firstStage("PICADO", "RES", bullGrade, random);

        //D[2] in the original
        gameState.toreadoresDamage = firstStage("TOREAD", "ORES", bullGrade, random);

        boolean done = false; //controls the main game loop

        while (!done) {

            gameState.passNumber++;
            System.out.printf("\n\nPASS NUMBER %d \n", gameState.passNumber);

            if (gameState.passNumber < MAX_PASSES_BEFORE_CHARGE) {
                System.out.println("THE BULL IS CHARGING AT YOU!  YOU ARE THE MATADOR--");
                System.out.print("DO YOU WANT TO KILL THE BULL? ");
            } else {
                System.out.print("HERE COMES THE BULL.  TRY FOR A KILL? ");
            }

            BooleanAnswer yesOrNo = readYesOrNo(scanner);
            if (yesOrNo.equals(BooleanAnswer.YES)) {
                gameState = attemptKillBull(bullGrade, random, gameState, scanner);
                done = true;
            } else {
                int capeMove;
                if (gameState.passNumber < MAX_PASSES_BEFORE_CHARGE) {
                    capeMove = readCapeMove("WHAT MOVE DO YOU MAKE WITH THE CAPE", scanner);
                } else {
                    capeMove = readCapeMove("CAPE MOVE", scanner);
                }

                //handle cape move
                gameState = handleCapeMove(capeMove, random, scanner, bullGrade, gameState);

                if (gameState.matadorStatus.equals(MatadorStatus.DEFEATED) || gameState.matadorStatus.equals(MatadorStatus.DEAD)) {
                    done = true;
                }
            }

        }

        crowdReaction(gameState, bullGrade, random);

        System.out.println("\nADIOS\n\n");
    }

    private static void printBullGradeInfo(BullGrade bullGrade) {
        System.out.println("\n\nYOU HAVE DRAWN A " + bullGrade.name() + " BULL.");
        if (bullGrade.equals(BullGrade.AWFUL)) {
            System.out.println("YOU'RE LUCKY.");
        } else if (bullGrade.equals(BullGrade.SUPERB)) {
            System.out.println("GOOD LUCK.  YOU'LL NEED IT.");
        }
    }

    private static void crowdReaction(GameState gameState, BullGrade bullGrade, Random random) {
        System.out.println("\n\n");
        if (!gameState.matadorStatus.equals(MatadorStatus.DEFEATED)) {
            if (!gameState.matadorStatus.equals(MatadorStatus.INJURED)) {
                if (gameState.bullStatus.equals(BullStatus.DEAD)) {
                    System.out.println("THE CROWD CHEERS!");
                }
            } else {
                System.out.println("THE CROWD CHEERS WILDLY!");
            }

            System.out.println("\nTHE CROWD AWARDS YOU");
            var crowdReactionScore = calculateCrowdReactionScore(gameState, bullGrade, random);
            if (crowdReactionScore < 2.4) {
                System.out.println("NOTHING AT ALL.");
            } else if (crowdReactionScore < 4.9) {
                System.out.println("ONE EAR OF THE BULL.");
            } else if (crowdReactionScore < 7.4) {
                System.out.println("BOTH EARS OF THE BULL!");
                System.out.println("OLE!");
            } else {
                System.out.println("OLE!  YOU ARE 'MUY HOMBRE'!! OLE!  OLE!");
            }

        } else {
            System.out.println("THE CROWD BOOS FOR TEN MINUTES.  IF YOU EVER DARE TO SHOW");
            System.out.println("YOUR FACE IN A RING AGAIN, THEY SWEAR THEY WILL KILL YOU--");
            System.out.println("UNLESS THE BULL DOES FIRST.");
        }
    }

    private static GameState handleCapeMove(int capeMove, Random random, Scanner scanner, BullGrade bullGrade, GameState gameState) {
        double m;
        if (capeMove == 0) {
            m = 3;
        } else if (capeMove == 1) {
            m = 2;
        } else {
            //capeMove == 2
            m = 0.5;
        }
        gameState.capeMovesCumulative += m;

        double f = (6 - bullGrade.getValue() + m / 10) * random.nextFloat() / ((gameState.picadoresDamage + gameState.toreadoresDamage + gameState.passNumber / 10d) * 5);
        if (f >= 0.51) {
            System.out.println("THE BULL HAS GORED YOU!");
            gameState = stateAfterGoring(random, scanner, gameState, bullGrade);
        }
        return gameState;
    }

    private static GameState stateAfterGoring(Random random, Scanner scanner, GameState gameState, BullGrade bullGrade) {
        GameState newGameState = gameState.newCopy();
        if (random.nextBoolean()) {
            System.out.println("YOU ARE DEAD.");
            newGameState.matadorStatus = MatadorStatus.DEAD;
        } else {
            System.out.println("YOU ARE STILL ALIVE.");
            newGameState = readAndHandleForfeitDecision(random, scanner, newGameState, bullGrade);
        }
        return newGameState;
    }

    /**
     * Calculate the crowd's reaction score based on the game state plus some randomness.
     * (FNC in the original code on line 1390)
     */
    private static double calculateCrowdReactionScore(GameState gameState, BullGrade bullGrade, Random random) {
        return calculateGameScore(gameState, bullGrade) * random.nextFloat();
    }

    /**
     * Calculates the ame score based on the current state and the selected bull grade.
     * (FND in the original code on line 1395)
     */
    private static double calculateGameScore(GameState gameState, BullGrade bullGrade) {
        return (4.5 + gameState.capeMovesCumulative / 6 - (gameState.picadoresDamage + gameState.toreadoresDamage) * 2.5 + 4 * gameState.matadorStatus.getValue() + 2 * gameState.bullStatus.getValue() - Math.pow(gameState.passNumber, 2) / 120f - bullGrade.getValue());
    }

    private static int readInt(String prompt, Scanner scanner) {
        System.out.print(prompt);
        while (true) {
            System.out.print("? ");
            String input = scanner.nextLine();
            try {
                return Integer.parseInt(input);
            } catch (NumberFormatException e) {
                System.out.println("!NUMBER EXPECTED - RETRY INPUT LINE");
            }
        }
    }

    private static int readCapeMove(String initialPrompt, Scanner scanner) {
        String prompt = initialPrompt;
        while (true) {
            int capeMove = readInt(prompt, scanner);
            if (capeMove <= 0 || capeMove > 3) {
                System.out.println("DON'T PANIC, YOU IDIOT!  PUT DOWN A CORRECT NUMBER");
                prompt = "";
            } else {
                return capeMove;
            }
        }
    }

    private static BooleanAnswer readYesOrNo(Scanner scanner) {
        while (true) {
            String answer = scanner.nextLine();
            if (answer.equalsIgnoreCase("YES")) {
                return BooleanAnswer.YES;
            } else if (answer.equalsIgnoreCase("NO")) {
                return BooleanAnswer.NO;
            } else {
                System.out.println("INCORRECT ANSWER - - PLEASE TYPE 'YES' OR 'NO'.");
            }
        }
    }

    private static void printInstructions() {
        System.out.print("HELLO, ALL YOU BLOODLOVERS AND AFICIONADOS.\n");
        System.out.print("HERE IS YOUR BIG CHANCE TO KILL A BULL.\n\n");
        System.out.print("ON EACH PASS OF THE BULL, YOU MAY TRY\n");
        System.out.print("0 - VERONICA (DANGEROUS INSIDE MOVE OF THE CAPE)\n");
        System.out.print("1 - LESS DANGEROUS OUTSIDE MOVE OF THE CAPE\n");
        System.out.print("2 - ORDINARY SWIRL OF THE CAPE.\n\n");
        System.out.print("INSTEAD OF THE ABOVE, YOU MAY TRY TO KILL THE BULL\n");
        System.out.print("ON ANY TURN: 4 (OVER THE HORNS), 5 (IN THE CHEST).\n");
        System.out.print("BUT IF I WERE YOU,\n");
        System.out.print("I WOULDN'T TRY IT BEFORE THE SEVENTH PASS.\n\n");
        System.out.print("THE CROWD WILL DETERMINE WHAT AWARD YOU DESERVE\n");
        System.out.print("(POSTHUMOUSLY IF NECESSARY).\n");
        System.out.print("THE BRAVER YOU ARE, THE BETTER THE AWARD YOU RECEIVE.\n\n");
        System.out.print("THE BETTER THE JOB THE PICADORES AND TOREADORES DO,\n");
        System.out.print("THE BETTER YOUR CHANCES ARE.\n\n");
    }

    private static void printHeader() {
        System.out.println("                                               BULL");
        System.out.println("                            CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    }

    /**
     * Random number 1 or 2
     * FNA in the original
     */
    private static int randomNumber1or2(Random random) {
        return (int) (random.nextFloat() * 2 + 1);
    }

    private static double firstStage(String horseType, String suffix, BullGrade bullGrade, Random random) {
        double b = ((3d / bullGrade.getValue()) * random.nextFloat());
        double c;

        if (b < 0.37) {
            c = 0.5;
        } else if (b < 0.5) {
            c = 0.4;
        } else if (b < 0.63) {
            c = 0.3;
        } else if (b < 0.87) {
            c = 0.2;
        } else {
            c = 0.1;
        }

        int t = (int) (10 * c + 0.2);

        System.out.println("THE " + horseType + suffix + " DID A " + getPerformanceRating(t) + " JOB.");

        if (t >= 4) {
            if (t == 5) {
                if (!horseType.equals("TOREAD")) {
                    System.out.println(" " + randomNumber1or2(random) + " OF THE HORSES OF THE " + horseType + suffix + " KILLED.");
                }
                System.out.println(" " + randomNumber1or2(random) + " OF THE " + horseType + suffix + " KILLED.");

            } else {
                if (random.nextBoolean()) {
                    System.out.println("ONE OF THE " + horseType + " " + suffix + " WAS KILLED.");
                } else {
                    System.out.println("NO " + horseType + " " + suffix + " WERE KILLED.");
                }
            }
        }

        System.out.println();
        return c;
    }

    public static String getPerformanceRating(int t) {
        switch (t) {
            case 1:
                return "SUPERB";
            case 2:
                return "GOOD";
            case 3:
                return "FAIR";
            case 4:
                return "POOR";
            default:
                return "AWFUL";
        }
    }

    private static GameState attemptKillBull(BullGrade bullGrade, Random random, GameState gameState, Scanner scanner) {
        GameState newGameState = gameState.newCopy();
        System.out.println("\nIT IS THE MOMENT OF TRUTH.\n");

        int h = readInt("HOW DO YOU TRY TO KILL THE BULL", scanner);
        if (h == 4 || h == 5) {
            var K = (6 - bullGrade.getValue()) * 10 * random.nextFloat() / ((gameState.picadoresDamage + gameState.toreadoresDamage) * 5 * gameState.passNumber);
            if (h == 4) {
                if (K > 0.8) {
                    System.out.println("THE BULL HAS GORED YOU!");
                    newGameState = stateAfterGoring(random, scanner, newGameState, bullGrade);
                } else {
                    System.out.println("YOU KILLED THE BULL!");
                    newGameState.bullStatus = BullStatus.DEAD;
                    return newGameState;//game over
                }
            } else {
                if (K > 0.2) {
                    System.out.println("THE BULL HAS GORED YOU!");
                    newGameState = stateAfterGoring(random, scanner, newGameState, bullGrade);
                } else {
                    System.out.println("YOU KILLED THE BULL!");
                    newGameState.bullStatus = BullStatus.DEAD;
                    return newGameState;//game over
                }
            }

        } else {
            System.out.println("YOU PANICKED.  THE BULL GORED YOU.");
            if (random.nextBoolean()) {
                System.out.println("YOU ARE DEAD.");
                newGameState.matadorStatus = MatadorStatus.DEAD;
                return newGameState;
            } else {
                return readAndHandleForfeitDecision(random, scanner, newGameState, bullGrade);
            }

        }

        return newGameState;
    }

    private static GameState readAndHandleForfeitDecision(Random random, Scanner scanner, GameState gameState, BullGrade bullGrade) {
        GameState newGameState = gameState.newCopy();

        System.out.print("\nDO YOU RUN FROM THE RING? ");
        BooleanAnswer yesOrNo = readYesOrNo(scanner);
        if (yesOrNo == BooleanAnswer.NO) {
            System.out.println("\n\nYOU ARE BRAVE.  STUPID, BUT BRAVE.");
            if (random.nextBoolean()) {
                newGameState.matadorStatus = MatadorStatus.INJURED;
                return newGameState;
            } else {
                System.out.println("YOU ARE GORED AGAIN!");
                return stateAfterGoring(random, scanner, newGameState, bullGrade);
            }
        } else {
            System.out.println("COWARD");
            newGameState.matadorStatus = MatadorStatus.DEFEATED;
            return newGameState;

        }
    }

    private static boolean readInstructionsChoice(Scanner scan) {
        System.out.print("DO YOU WANT INSTRUCTIONS? ");

        String choice = scan.nextLine();
        return !choice.equalsIgnoreCase("NO");
    }


    private enum BooleanAnswer {
        YES, NO
    }

    private enum BullGrade {
        SUPERB(1),
        GOOD(2),
        FAIR(3),
        POOR(4),
        AWFUL(5);

        private final int value;

        BullGrade(int value) {
            this.value = value;
        }

        public int getValue() {
            return value;
        }

        public static BullGrade fromValue(int value) {
            for (BullGrade grade : BullGrade.values()) {
                if (grade.getValue() == value) {
                    return grade;
                }
            }
            throw new IllegalArgumentException("Invalid value: " + value);
        }

    }

    /**
     * Represents the game state.
     */
    private static class GameState {
        private MatadorStatus matadorStatus; //D[4] in the original
        private BullStatus bullStatus; //D[5] in the original
        private double picadoresDamage; //D[1] in the original
        private double toreadoresDamage; //D[2] in the original
        private int passNumber; //D[3] in the original
        private double capeMovesCumulative; //L in the original

        public GameState() {
            picadoresDamage = 0;
            toreadoresDamage = 0;
            passNumber = 0;
            matadorStatus = MatadorStatus.ALIVE;
            bullStatus = BullStatus.ALIVE;
            capeMovesCumulative = 1;
        }

        public GameState newCopy() {
            GameState newState = new GameState();
            newState.matadorStatus = this.matadorStatus;
            newState.bullStatus = this.bullStatus;
            newState.picadoresDamage = this.picadoresDamage;
            newState.toreadoresDamage = this.toreadoresDamage;
            newState.passNumber = this.passNumber;
            newState.capeMovesCumulative = this.capeMovesCumulative;
            return newState;
        }
    }

    private enum MatadorStatus {
        ALIVE(1),
        INJURED(2),
        DEAD(1.5),
        DEFEATED(0);

        private final double value;

        MatadorStatus(double value) {
            this.value = value;
        }

        public double getValue() {
            return value;
        }
    }

    private enum BullStatus {
        ALIVE(1),
        DEAD(2);

        private final double value;

        BullStatus(double value) {
            this.value = value;
        }

        public double getValue() {
            return value;
        }
    }

}

